packages/utils/node-resolver-core/src/Wrapper.js

```diff
@@ -0,0 +1,764 @@
+          options.env,
+          this.options.mode,
+        ),
+        moduleDirResolver:
+          process.versions.pnp != null
+            ? (module, from) => {
+                // $FlowFixMe[prop-missing]
+                let pnp = _Module.findPnpApi(path.dirname(from));
+
+                return pnp.resolveToUnqualified(
+                  // append slash to force loading builtins from npm
+                  module + '/',
+                  from,
+                );
+              }
+            : undefined,
+      });
+      this.resolversByEnv.set(options.env.id, resolver);
+    }
+
+    // Special case for entries. Convert absolute paths to relative from project root.
+    if (options.parent == null) {
+      options.parent = path.join(this.options.projectRoot, 'index');
+      if (path.isAbsolute(options.filename)) {
+        options.filename = relativePath(
+          this.options.projectRoot,
+          options.filename,
+        );
+      }
+    }
+
+    let res = resolver.resolve(options);
+
+    // Invalidate whenever the .pnp.js file changes.
+    // TODO: only when we actually resolve a node_modules package?
+    if (
+      process.versions.pnp != null &&
+      options.parent &&
+      res.invalidateOnFileChange
+    ) {
+      // $FlowFixMe[prop-missing]
+      let pnp = _Module.findPnpApi(path.dirname(options.parent));
+      res.invalidateOnFileChange.push(pnp.resolveToUnqualified('pnpapi', null));
+    }
+
+    if (res.error) {
+      let diagnostic = await this.handleError(res.error, options);
+      return {
+        diagnostics: Array.isArray(diagnostic)
+          ? diagnostic
+          : diagnostic
+          ? [diagnostic]
+          : [],
+        invalidateOnFileCreate: res.invalidateOnFileCreate,
+        invalidateOnFileChange: res.invalidateOnFileChange,
+      };
+    }
+
+    switch (res.resolution?.type) {
+      case 'Path':
+        return {
+          filePath: res.resolution.value,
+          invalidateOnFileCreate: res.invalidateOnFileCreate,
+          invalidateOnFileChange: res.invalidateOnFileChange,
+          sideEffects: res.sideEffects,
+          query: res.query != null ? new URLSearchParams(res.query) : undefined,
+        };
+      case 'Builtin':
+        return this.resolveBuiltin(res.resolution.value, options);
+      case 'External': {
+        if (
+          options.sourcePath &&
+          options.env.isLibrary &&
+          options.specifierType !== 'url'
+        ) {
+          let diagnostic = await this.checkExcludedDependency(
+            options.sourcePath,
+            options.filename,
+            options,
+          );
+          if (diagnostic) {
+            return {
+              diagnostics: [diagnostic],
+              invalidateOnFileCreate: res.invalidateOnFileCreate,
+              invalidateOnFileChange: res.invalidateOnFileChange,
+            };
+          }
+        }
+
+        // TODO: invalidations?
+        return {isExcluded: true};
```

Comment:
Changing aliases/exports/... could probably influence whether it's external or not?